plugins/CoreVisualizations/Visualizations/Graph.php

```diff
@@ -135,26 +149,48 @@ public function afterAllFiltersAreApplied()
             ));
         }
 
-        // set default selectable columns, if none specified
-        $selectableColumns = $this->config->selectable_columns;
-        if (false === $selectableColumns) {
-            $selectableColumns = array('nb_visits', 'nb_actions', 'nb_uniq_visitors', 'nb_users');
-
-            if ($this->config->show_goals) {
-                $goalMetrics       = array('nb_conversions', 'revenue');
-                $selectableColumns = array_merge($selectableColumns, $goalMetrics);
-            }
-        }
-
         $transformed = array();
-        foreach ($selectableColumns as $column) {
+        foreach ($this->config->selectable_columns as $column) {
             $transformed[] = array(
                 'column'      => $column,
                 'translation' => @$this->config->translations[$column],
                 'displayed'   => in_array($column, $this->config->columns_to_display)
             );
         }
-
         $this->config->selectable_columns = $transformed;
     }
+
+    protected function generateSelectableColumns()
+    {
+        $selectableColumns = array('nb_visits', 'nb_actions');
+
+        // Only add unique visitors and users metrics when there is data to support them
+        $firstRow = $this->getDataTable()->getFirstRow();
+        if (!$firstRow || $firstRow->hasColumn('nb_uniq_visitors')) {
+            $selectableColumns[] = 'nb_uniq_visitors';
+        }
+        if (!$firstRow || $firstRow->hasColumn('nb_users')) {
+            $selectableColumns[] = 'nb_users';
+        }
+
+        if ($this->config->show_goals) {
+            $goalMetrics       = array('nb_conversions', 'revenue');
+            $selectableColumns = array_merge($selectableColumns, $goalMetrics);
+        }
+
+        $this->config->selectable_columns = $selectableColumns;
+    }
+
+    protected function ensureValidColumnsToDisplay()
+    {
+        $columnsToDisplay = $this->config->columns_to_display;
```

Comment:
I wonder if this logic here is needed? I wonder about possible side effects where reports declare `columns_to_display` but not `selectable_columns`.

In general this is a great addition that makes sense though. If something is not selectable, we shouldn't show it. Just scared about possible side effects or breaking something. Might not be an issue though but have a feeling there are some reports that might break as I can't find where the reports usually add this info.

Reply:
We can quickly chat about it tmrw might be easiest

Reply:
I've changed it to match against `$dataTable->getColumns()` rather than `selectable_columns` - this should hopefully avoid any side effects.